### PR TITLE
add option to `compilecache` so that Pkg precompilation can not skip different versions of loaded modules

### PR DESCRIPTION
**Pull request by IanButterworth**
Paired with https://github.com/JuliaLang/Pkg.jl/pull/2484

If I understand correctly, classical code-load precompilation skips precompiling different versions of loaded packages to save time, given they won't be loadable in the current session.

For the new pkg precompilation process this seems less critical.

The `compilecache` option added in this PR, along with https://github.com/JuliaLang/Pkg.jl/pull/2484 allows Pkg precompilation to precompile the different version of loaded packages, but warn the user that the package is loaded so the new version won't be accessible until after a session restart.

This avoids the `?` state in pkg precompilation that delayed precompilation until after session restart, which means the remaining precompilation was likely to happen at slower code-load as the user has to remember to go back to pkg mode and precompile.

The new behavior does all the precompilation in the current session, but reports as yellow `✓`'s, with an explanation that a restart is needed to access the new versions.

More details in https://github.com/JuliaLang/Pkg.jl/pull/2484

### CODE REVIEW

**Comment by vtjnash** [comment]
Might need a better (more descriptive) name, such as `ignore_loaded_modules`, but seems good otherwise

**Comment by vtjnash** [comment]
Or could take `concrete_deps` as an argument itself (eventually, this is supposed to take a single environment, but we currently don't have a good way to reify those across processes)

**Comment by IanButterworth** [comment]
I assume the sockets test failure is unrelated. Can we merge this so the Pkg PR can be tested?

**Comment by IanButterworth** [comment]
Btw, the first suggestion was easier so I went with that. Thanks

File: base/loading.jl [hunk 1]
```diff
@@ -1303,16 +1303,20 @@ end
 
 const MAX_NUM_PRECOMPILE_FILES = Ref(10)
 
-function compilecache(pkg::PkgId, path::String, internal_stderr::IO = stderr, internal_stdout::IO = stdout)
+function compilecache(pkg::PkgId, path::String, internal_stderr::IO = stderr, internal_stdout::IO = stdout,
+        ignore_loaded_modules::Bool = true)
```
Comment (staticfloat) [inline base/loading.jl:1307]: Can you add enough spaces to match this up with the rest of the args?

Comment (IanButterworth) [reply base/loading.jl]: yep! done

**Pull request merged**